homeassistant/components/ecoforest/switch.py

```diff
@@ -0,0 +1,75 @@
+"""Switch platform for Ecoforest."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from pyecoforest.models.device import Device
+
+from homeassistant.components.switch import SwitchEntity, SwitchEntityDescription
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import EcoforestCoordinator
+from .entity import EcoforestEntity
+
+
+@dataclass
+class EcoforestRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[Device], bool]
+
+
+@dataclass
+class EcoforestSwitchEntityDescription(
+    SwitchEntityDescription, EcoforestRequiredKeysMixin
+):
+    """Describes an Ecoforest switch entity."""
+
+
+SWITCH_TYPES: tuple[EcoforestSwitchEntityDescription, ...] = (
+    EcoforestSwitchEntityDescription(
+        key="status", translation_key="status", value_fn=lambda data: data.on
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Ecoforest switch platform."""
+    coordinator: EcoforestCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+
+    entities = [
+        EcoforestSwitchEntity(coordinator, description) for description in SWITCH_TYPES
+    ]
+
+    async_add_entities(entities)
+
+
+class EcoforestSwitchEntity(EcoforestEntity, SwitchEntity):
+    """Representation of an Ecoforest switch entity."""
+
+    entity_description: EcoforestSwitchEntityDescription
+
+    _attr_name = None
+
+    @property
+    def is_on(self) -> bool:
+        """Return the state of the ecoforest device."""
+        return self.entity_description.value_fn(self.data)
+
+    async def async_turn_on(self):
+        """Turn on the ecoforest device."""
+        await self.coordinator.api.turn(True)
+        await self.coordinator.async_request_refresh()
+
+    async def async_turn_off(self):
+        """Turn off the ecoforest device."""
+        await self.coordinator.api.turn(False)
+        await self.coordinator.async_request_refresh()
```

Comment:
Is this going to be the only switch? Currently you use an entity description, but the way to turn it on and off is the same for every switch. To fully utilize entity descriptions, please create 2 functions like the value_fn like on_fn and off_fn which you execute in here.

Reply:
I considered it, but since this is likely the only switch, but I will change so that get's ready.

Reply:
Just trying to figure out to make the lambda work with async function

Reply:
`Callable[[Device], Awaitable[None]]`

Reply:
I have used `switch_fn: Callable[[EcoforestApi, bool], Coroutine[Any, Any, Device]]`

Reply:
didnt awaitable work?

Reply:
I wasn't able to make it work, to be honest not sure why.

Reply:
I guess i know why, it's because the api returns a device so it needs to be `Awaitable[Device]` instead

Reply:
Can you try that in here?

Reply:
I was already creating a commit for it. Already there.